olive/passes/pytorch/qlora.py

```diff
@@ -0,0 +1,487 @@
+        new_model_loading_args = model_loading_args.dict()
+        new_model_loading_args.pop("device_map")
+        # get new hf_config, will reset the model_loading_args and components
+        new_hf_config = model.hf_config.dict()
+        new_hf_config.pop("components")
+        new_hf_config["model_loading_args"] = new_model_loading_args
+
+        # get model path if any from the input model
+        # will access the model_path resource directly in case it is not local
+        model_path = model.resource_paths["model_path"]
+
+        # the output model is the same transformer pytorch model, but with the adapter_path and
+        # model_loading_args updated
+        return PyTorchModel(model_path=model_path, hf_config=new_hf_config, adapter_path=adapter_path)
+
+    @classmethod
+    def get_model_tokenizer(cls, model: PyTorchModel, config: ConfigBase) -> torch.nn.Module:
+        from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+        from peft.tuners.lora import LoraLayer
+        from transformers import AutoTokenizer
+
+        if not model.hf_config:
+            raise ValueError("QLoRA pass only supports PyTorchModel with hf_config.")
+
+        model_name = model.hf_config.model_name
+        model_path = model.model_path  # can be None, if so, model_name is used to load model
+        task = model.hf_config.task
+
+        # get peft task type
+        peft_task_type = get_peft_task_type_from_task(task, fail_on_not_found=True)
+
+        # compute_dtype
+        supported_dtypes = {
+            "bfloat16": torch.bfloat16,
+            "float16": torch.float16,
+            "float32": torch.float32,
+        }
+        assert (
+            config.compute_dtype in supported_dtypes
+        ), f"compute_dtype must be one of {list(supported_dtypes.keys())} but got {config['compute_dtype']}"
+        compute_dtype = supported_dtypes[config.compute_dtype]
+
+        # load model
+        if model.hf_config.model_loading_args:
+            logger.warning(
+                "Input model has model_loading_args. Ignoring. QLoRA will use its own model_loading_args based on the"
+                " pass config."
+            )
+        model_loading_args = HFModelLoadingArgs(
+            torch_dtype=compute_dtype,
+            # TODO: Worry about `use_multi_gpu` and distributed training later
+            # this uses all available GPUs, model parallel
+            device_map="auto",
+            quantization_method="bitsandbytes",
+            quantization_config={
+                "load_in_4bit": True,
+                "bnb_4bit_compute_dtype": compute_dtype,
+                "bnb_4bit_use_double_quant": config.double_quant,
+                "bnb_4bit_quant_type": config.quant_type,
+            },
+        )
+        pytorch_model = load_huggingface_model_from_task(
+            task=task, name=model_path or model_name, **model_loading_args.get_loading_args()
+        )
```

Comment:
could I suppose the pytorch_model is same with model.load_model?

Reply:
It is different here since the pass provides it's own model_loading_args. That's why it gives the warning above this line if the input model had it's own model_loading_args.

Reply:
we could assign the new model_load_args to the input model and call the load_model again?

Reply:
That is possible if we deepcopy the input model and assign the new model loading args. since we don't want to update the input model object or it's loaded model if any. But I think, that's more complicated?

Reply:
I suppose we could directly return the deepcopy model as the pass output. In line 280, we construct everything and copy the original model stuff for the new output model. I think that's is same with the deepcopy. 

If we directly use deepcopy, all of the construction stuff could be removed and code will look simpler and duplicated could be removed?

Reply:
the only thing we need to do before return the deepcopy model is clean the cached pytorch model (PyTorchModel.model)?

Reply:
That sounds like a good idea. 
Just pushed a commit with this change